test/logging/logging.go

```diff
@@ -64,11 +70,21 @@ func (e *zapMetricExporter) ExportView(vd *view.Data) {
 	e.logger.Debug(spew.Sprint(vd))
 }
 
-// ExportSpan will emit the trace data to the zap logger.
+// GetEmitableSpan starts and returns a trace.Span with a name that
+// is used by the ExportSpan method to emit the span.
+func GetEmitableSpan(ctx context.Context, metricName string) *trace.Span {
+	_, span := trace.StartSpan(ctx, emitableSpanNamePrefix + metricName)
+	return span
+}
+
+// ExportSpan will emit the trace data to the zap logger. The span is emitted
+// only if the metric name is prefix with emitableSpanNamePrefix constant.
 func (e *zapMetricExporter) ExportSpan(vd *trace.SpanData) {
-	duration := vd.EndTime.Sub(vd.StartTime)
-	// We will start the log entry with `metric` to identify it as a metric for parsing
-	e.logger.Infof("metric %s %d %d %s", vd.Name, vd.StartTime.UnixNano(), vd.EndTime.UnixNano(), duration)
+	if strings.HasPrefix(vd.Name, emitableSpanNamePrefix) {
+		duration := vd.EndTime.Sub(vd.StartTime)
+		// We will start the log entry with `metric` to identify it as a metric for parsing
+		e.logger.Infof("metric %s %d %d %s", vd.Name[len(emitableSpanNamePrefix):], vd.StartTime.UnixNano(), vd.EndTime.UnixNano(), duration)
+	}
```

Comment:
do we want to throw an eror or atleast log a message if someone calls ExportSpan() without the prefix?

Reply:
That would result in what currently gets showed up metric logs replaced with these error messages. Polluting the logs even more.

Reply:
I dont think adding a log stmt will affect this. the problem is we have multiple statements from tracing and us that are of form `metric foo 1:2` and `metric /bar 3:4`. The parser is looking for all log statements that have this form and ignores the rest. So, having a lot statement that simple says `Could not export span as foo did not have prefix` will not show up in metric logs, but will be in the build logs, if we want to look for missing spans.

Reply:
I got what the initial problem is. I am saying that adding the log statement of a missing span is not of much use. Also it would result in unnecessary log statements when we enable `trace.AlwaysSample()` which is needed for zipkin tracing. As an example take a look at the logs in the attached file that would be generated for a autoscale test `TestAutoscaleUpCountPods`.  Lines like these would be noise in logs rather than adding any value
```
2019-02-15T13:05:37.886-0800    info    SpoofUtil       trace/trace.go:266      Not logging Span : 
2019-02-15T13:05:37.886-0800    info    SpoofUtil       trace/trace.go:266      Not logging Span : SpoofingClient-Trace
```

Reply:
Attachment is failing, addding a snippet instead, will email you if you need the whole log file.

```
ok      github.com/knative/serving/test/conformance     0.188s [no tests to run]
2019-02-15T13:05:28.783-0800    info    logging/config.go:39    Successfully created the logger.        {"knative.dev/jsonconfig": "{\n\t  \"level\": \"info\",\n\t  \"encoding\": \"console\",\n\t  \"outputPaths\": [\"stdout\"],\n\t  \"errorOutputPaths\": [\"stderr\"],\n\t  \"encoderConfig\": {\n\t    \"timeKey\": \"ts\",\n\t    \"messageKey\": \"message\",\n\t    \"levelKey\": \"level\",\n\t    \"nameKey\": \"logger\",\n\t    \"callerKey\": \"caller\",\n\t    \"messageKey\": \"msg\",\n\t    \"stacktraceKey\": \"stacktrace\",\n\t    \"lineEnding\": \"\",\n\t    \"levelEncoder\": \"\",\n\t    \"timeEncoder\": \"iso8601\",\n\t    \"durationEncoder\": \"\",\n\t    \"callerEncoder\": \"\"\n\t  }\n\t}"}
2019-02-15T13:05:28.783-0800    info    logging/config.go:39    Logging level set to info
2019-02-15T13:05:28.783-0800    warn    logging/config.go:41    Fetch GitHub commit ID from kodata failed: "KO_DATA_PATH" does not exist or is empty
=== RUN   TestAutoscaleUpDownUp
2019-02-15T13:05:29.501-0800    info    SpoofUtil       zipkin/util.go:78       Zipkin port-forward process started with PID: 150744
2019-02-15T13:05:29.501-0800    info    SpoofUtil       zipkin/util.go:83       Successfully setup SpoofingClient for Zipkin Tracing
2019-02-15T13:05:29.502-0800    info    TestAutoscaleUpDownUp   e2e/autoscale_test.go:148       Creating a new Route and Configuration
2019-02-15T13:05:29.503-0800    info    TestAutoscaleUpDownUp   test/crd.go:280 Seeding rand.Rand with 1550264729503005400
2019-02-15T13:05:29.769-0800    info    TestAutoscaleUpDownUp   e2e/autoscale_test.go:157       When the Revision can have traffic routed to it, the Route is marked as Ready.
2019-02-15T13:05:35.792-0800    info    SpoofUtil       trace/trace.go:266      metric WaitForRouteState/noodleburgtwhlqbhx/RouteIsReady 1550264729769391297 1550264735792242666 6.022851369s
2019-02-15T13:05:35.792-0800    info    TestAutoscaleUpDownUp   e2e/autoscale_test.go:167       Serves the expected data at the endpoint
2019-02-15T13:05:36.849-0800    info    SpoofUtil       trace/trace.go:266      Not logging Span : 
2019-02-15T13:05:36.849-0800    info    SpoofUtil       trace/trace.go:266      Not logging Span : SpoofingClient-Trace
2019-02-15T13:05:36.849-0800    info    SpoofUtil       trace/trace.go:266      metric WaitForEndpointState/CheckingEndpointAfterUpdating 1550264735812719071 1550264736849812661 1.03709359s
2019-02-15T13:05:36.849-0800    info    TestAutoscaleUpDownUp   e2e/autoscale_test.go:205       The autoscaler spins up additional replicas when traffic increases.
2019-02-15T13:05:36.849-0800    info    TestAutoscaleUpDownUp   e2e/autoscale_test.go:70        Maintaining 20 concurrent requests for 20s.
2019-02-15T13:05:36.849-0800    info    TestAutoscaleUpDownUp   e2e/autoscale_test.go:114       Waiting for all requests to complete.
2019-02-15T13:05:37.886-0800    info    SpoofUtil       trace/trace.go:266      Not logging Span : 
2019-02-15T13:05:37.886-0800    info    SpoofUtil       trace/trace.go:266      Not logging Span : SpoofingClient-Trace
2019-02-15T13:05:37.886-0800    info    TestAutoscaleUpDownUp   e2e/autoscale_test.go:98        Request took: 1.01221886s
2019-02-15T13:05:37.899-0800    info    SpoofUtil       trace/trace.go:266      Not logging Span : 
2019-02-15T13:05:37.899-0800    info    SpoofUtil       trace/trace.go:266      Not logging Span : SpoofingClient-Trace
2019-02-15T13:05:37.899-0800    info    TestAutoscaleUpDownUp   e2e/autoscale_test.go:98        Request took: 1.022484222s
2019-02-15T13:05:37.900-0800    info    SpoofUtil       trace/trace.go:266      Not logging Span : 
2019-02-15T13:05:37.900-0800    info    SpoofUtil       trace/trace.go:266      Not logging Span : SpoofingClient-Trace
2019-02-15T13:05:37.900-0800    info    TestAutoscaleUpDownUp   e2e/autoscale_test.go:98        Request took: 1.024807069s
2019-02-15T13:05:37.900-0800    info    SpoofUtil       trace/trace.go:266      Not logging Span : 
2019-02-15T13:05:37.900-0800    info    SpoofUtil       trace/trace.go:266      Not logging Span : SpoofingClient-Trace
2019-02-15T13:05:37.900-0800    info    TestAutoscaleUpDownUp   e2e/autoscale_test.go:98        Request took: 1.023475925s
2019-02-15T13:05:37.905-0800    info    SpoofUtil       trace/trace.go:266      Not logging Span : 
2019-02-15T13:05:37.905-0800    info    SpoofUtil       trace/trace.go:266      Not logging Span : SpoofingClient-Trace
2019-02-15T13:05:37.905-0800    info    TestAutoscaleUpDownUp   e2e/autoscale_test.go:98        Request took: 1.030015209s
2019-02-15T13:05:37.905-0800    info    SpoofUtil       trace/trace.go:266      Not logging Span : 
2019-02-15T13:05:37.905-0800    info    SpoofUtil       trace/trace.go:266      Not logging Span : SpoofingClient-Trace
2019-02-15T13:05:37.905-0800    info    TestAutoscaleUpDownUp   e2e/autoscale_test.go:98        Request took: 1.031238232s
2019-02-15T13:05:37.918-0800    info    SpoofUtil       trace/trace.go:266      Not logging Span : 
2019-02-15T13:05:37.918-0800    info    SpoofUtil       trace/trace.go:266      Not logging Span : SpoofingClient-Trace
2019-02-15T13:05:37.918-0800    info    TestAutoscaleUpDownUp   e2e/autoscale_test.go:98        Request took: 1.042979217s
2019-02-15T13:05:37.921-0800    info    SpoofUtil       trace/trace.go:266      Not logging Span : 
2019-02-15T13:05:37.921-0800    info    SpoofUtil       trace/trace.go:266      Not logging Span : SpoofingClient-Trace
2019-02-15T13:05:37.921-0800    info    TestAutoscaleUpDownUp   e2e/autoscale_test.go:98        Request took: 1.044741757s
2019-02-15T13:05:37.925-0800    info    SpoofUtil       trace/trace.go:266      Not logging Span : 
2019-02-15T13:05:37.925-0800    info    SpoofUtil       trace/trace.go:266      Not logging Span : SpoofingClient-Trace
2019-02-15T13:05:37.925-0800    info    TestAutoscaleUpDownUp   e2e/autoscale_test.go:98        Request took: 1.050065696s
2019-02-15T13:05:37.969-0800    info    SpoofUtil       trace/trace.go:266      Not logging Span : 
2019-02-15T13:05:37.969-0800    info    SpoofUtil       trace/trace.go:266      Not logging Span : SpoofingClient-Trace
2019-02-15T13:05:37.969-0800    info    TestAutoscaleUpDownUp   e2e/autoscale_test.go:98        Request took: 1.095026009s
2019-02-15T13:05:38.887-0800    info    SpoofUtil       trace/trace.go:266      Not logging Span : 
2019-02-15T13:05:38.887-0800    info    SpoofUtil       trace/trace.go:266      Not logging Span : SpoofingClient-Trace
2019-02-15T13:05:38.887-0800    info    TestAutoscaleUpDownUp   e2e/autoscale_test.go:98        Request took: 1.826152704s
2019-02-15T13:05:38.901-0800    info    SpoofUtil       trace/trace.go:266      Not logging Span : 
2019-02-15T13:05:38.902-0800    info    SpoofUtil       trace/trace.go:266      Not logging Span : SpoofingClient-Trace
2019-02-15T13:05:38.902-0800    info    TestAutoscaleUpDownUp   e2e/autoscale_test.go:98        Request took: 1.441207785s
2019-02-15T13:05:38.902-0800    info    SpoofUtil       trace/trace.go:266      Not logging Span : 
2019-02-15T13:05:38.902-0800    info    SpoofUtil       trace/trace.go:266      Not logging Span : SpoofingClient-Trace
2019-02-15T13:05:38.902-0800    info    TestAutoscaleUpDownUp   e2e/autoscale_test.go:98        Request took: 1.240331138s
2019-02-15T13:05:38.902-0800    info    SpoofUtil       trace/trace.go:266      Not logging Span : 
2019-02-15T13:05:38.902-0800    info    SpoofUtil       trace/trace.go:266      Not logging Span : SpoofingClient-Trace
2019-02-15T13:05:38.902-0800    info    TestAutoscaleUpDownUp   e2e/autoscale_test.go:98        Request took: 1.640669342s
2019-02-15T13:05:38.912-0800    info    SpoofUtil       trace/trace.go:266      Not logging Span : 
2019-02-15T13:05:38.912-0800    info    SpoofUtil       trace/trace.go:266      Not logging Span : SpoofingClient-Trace
2019-02-15T13:05:38.912-0800    info    TestAutoscaleUpDownUp   e2e/autoscale_test.go:98        Request took: 1.051948664s
2019-02-15T13:05:38.923-0800    info    SpoofUtil       trace/trace.go:266      Not logging Span : 
2019-02-15T13:05:38.923-0800    info    SpoofUtil       trace/trace.go:266      Not logging Span : 
```

Reply:
SGTM. Thanks for explanation.